### DISCUSSION TITLE
CIP 30: Taproot Address Format

### DISCUSSION THREAD

**Discussion by Jpja** [CounterpartyXCP/CIPs#101] [CIPs]
Discuss [CIP30: Taproot Support](https://github.com/CounterpartyXCP/cips/blob/master/cip-0030.md) here. Please stay on topic and be respectful.

Is there anything in the proposal that is unclear or needs clarification?

Objections or concerns should be documented here. Even if you are positive to the CIP overall, it is encouraged to list negative tradeoffs.

As I understand, the CIP is already coded. Can @jdogresorg or @pataegrillo please post a link to the code for review?

I suggest one month to document pros and cons of the CIP, and then poll the community on whether to implement the CIP as soon as possible, put in on hold, or reject it.

**Comment by pataegrillo** [comment]
Hi, this is the link to the Pull Request with all [taproot changes in counterparty-lib ](https://github.com/CounterpartyXCP/counterparty-lib/pull/1236).

The changes for python-bitcoinlib are [here](https://github.com/petertodd/python-bitcoinlib/compare/master...pataegrillo:python-bitcoinlib:master)

Just to clarify: the file requirements.txt inside counterparty-lib now points to a fork of python-bitcoinlib on my repository with the changes to encode/decode bechm32 (taproot) addresses, before was pointing to a fork on John Villar's repository. Later, this lib has to be included inside counterparty-lib project

**Comment by jotapea** [comment]
My initial feedback from https://github.com/CounterpartyXCP/cips/pull/99#issuecomment-1636831596:

> Considering the CIP by itself, I am in the camp of not requiring the latest version of Bitcoin Core to run Counterparty. This proposal would mean that counterparty-lib will require Bitcoin Core v24 as the minimum version, right? At the moment the minimum version of Bitcoin required for counterparty-lib is lower, I'm running v0.21.1 but I think a couple of lower versions should still work.

> Finding these lowest compatible versions should be part of this conversation. I believe that Counterparty's strength is in its legacy and working with older versions of Bitcoin honors this legacy.

TLDR: We must learn the lowest version of Bitcoin Core that works with the current counterparty-lib, and compare it with the lowest version that would now be required for this feature.

**Reply by jotapea** [reply]
Are there any protocol features that won't work with taproot addresses?

**Comment by pataegrillo** [comment]
Hi, for me (and this is my personal opinion), being counterparty on top of bitcoin and depending on its operation and protocols (not like other libraries that can be substituted), counterparty should try to use the last bitcoin version. Please, don't be offended, but i'm not aware of the legacy you are mentioning, what i do know is bitcoin and counterparty have change a lot, and even with all those changes ppl still have their tokens on theirs addresses and they can trade them without any problem as they have been doing for the past almost 10 years. Also, counterparty has backward compatibility, you could parse the whole data and have the same results using an official counterparty version.

Effectively, I'm using bitcoin v24.0.0 to enable taproot addresses. I didn't find any protocol function that doesn't work. In fact, to do some minor taproot testing, I did some basic counterparty transactions without any problems. Note that I can't test my own code, and it's in the testing phase when broken features will surface and need to be fixed before even thinking about upgrading to production.

By the way, doing some tests with taproot addresses, some dispensers were triggered (in my version, of course, not the official), so ppl were trying to buy some tokens with taproot addresses and they lost their money. Even if the dispenser were triggered, they would not be able to move those tokens because, for counterparty, taproot addresses are unknown.
